File: src/reactive/reducer.py
"""
Reducer types for State Management System

Implementación de: VELA-577 (TASK-035S)
Sprint: Sprint 15
Fecha: 2025-12-02

Descripción:
Este módulo define los tipos para Reducers en el sistema de State Management.
Los Reducers son funciones puras: (State, Action) → State
"""

from typing import Callable, TypeVar, Any, Dict
from abc import ABC, abstractmethod
from .action import Action


S = TypeVar('S')  # State type
A = TypeVar('A', bound=Action)  # Action type


# Type alias para una función reducer
Reducer = Callable[[S, A], S]


class ReducerProtocol(ABC):
    """
    Protocolo abstracto para reducers.
    
    Los reducers DEBEN ser funciones puras:
    - No side effects
    - Mismos inputs → mismo output
    - No mutar el estado original
    
    Ejemplo:
        def counter_reducer(state: int, action: Action) -> int:
            if action.get_type() == "INCREMENT":
                return state + 1
            elif action.get_type() == "DECREMENT":
                return state - 1
            return state
    """
    
    @abstractmethod
    def reduce(self, state: S, action: A) -> S:
        """
        Procesa una acción y retorna el nuevo estado.
        
        Args:
            state: Estado actual
            action: Acción a procesar
            
        Returns:
            Nuevo estado (NUNCA mutar el estado original)
        """
        pass


class CombinedReducer(ReducerProtocol):
    """
    Combina múltiples reducers en uno solo.
    
    Cada reducer maneja una parte del estado global.
    Similar a combineReducers() de Redux.
    
    Ejemplo:
        todos_reducer = lambda state, action: state
        counter_reducer = lambda state, action: state
        
        combined = CombinedReducer({
            "todos": todos_reducer,
            "counter": counter_reducer
        })
        
        # State: { todos: [...], counter: 0 }
    """
    
    def __init__(self, reducers: Dict[str, Reducer]):
        """
        Inicializa el combined reducer.
        
        Args:
            reducers: Dict de {key: reducer_function}
                     Cada reducer maneja state[key]
        """
        self.reducers = reducers
    
    def reduce(self, state: Dict[str, Any], action: Action) -> Dict[str, Any]:
        """
        Aplica cada reducer a su parte del estado.
        
        Args:
            state: Estado completo (dict)
            action: Acción a procesar
            
        Returns:
            Nuevo estado con todas las partes actualizadas
        """
        new_state = {}
        state_changed = False
        
        for key, reducer in self.reducers.items():
            prev_state = state.get(key)
            next_state = reducer(prev_state, action)
            
            new_state[key] = next_state
            
            # Detectar cambios (shallow comparison)
            if next_state is not prev_state:
                state_changed = True
        
        # Si nada cambió, retornar el mismo objeto (optimización)
        return new_state if state_changed else state


class ReducerBuilder:
    """
    Builder para construir reducers de forma declarativa.
    
    Ejemplo:
        reducer = (
            ReducerBuilder(initial_state=0)
            .case("INCREMENT", lambda state, action: state + 1)
            .case("DECREMENT", lambda state, action: state - 1)
            .case("SET", lambda state, action: action.payload)
            .build()
        )
    """
    
    def __init__(self, initial_state: Any = None):
        """
        Inicializa el builder.
        
        Args:
            initial_state: Estado inicial opcional
        """
        self.initial_state = initial_state
        self.handlers: Dict[str, Reducer] = {}
        self.default_handler: Reducer | None = None
    
    def case(self, action_type: str, handler: Reducer):
        """
        Registra un handler para un tipo de acción.
        
        Args:
            action_type: Tipo de acción a manejar
            handler: Función (state, action) -> new_state
            
        Returns:
            self (para encadenamiento)
        """
        self.handlers[action_type] = handler
        return self
    
    def default(self, handler: Reducer):
        """
        Registra un handler por defecto para acciones no manejadas.
        
        Args:
            handler: Función (state, action) -> new_state
            
        Returns:
            self (para encadenamiento)
        """
        self.default_handler = handler
        return self
    
    def build(self) -> Reducer:
        """
        Construye el reducer final.
        
        Returns:
            Función reducer
        """
        handlers = self.handlers
        default_handler = self.default_handler
        
        def reducer(state: Any, action: Action) -> Any:
            action_type = action.get_type()
            
            # Buscar handler específico
            if action_type in handlers:
                return handlers[action_type](state, action)
            
            # Usar handler por defecto
            if default_handler:
                return default_handler(state, action)
            
            # Sin cambios
            return state
        
        return reducer


def create_reducer(handlers: Dict[str, Reducer], default: Reducer | None = None) -> Reducer:
    """
    Función helper para crear reducers con un dict de handlers.
    
    Args:
        handlers: Dict de {action_type: handler_function}
        default: Handler por defecto opcional
        
    Returns:
        Función reducer
        
    Ejemplo:
        reducer = create_reducer({
            "INCREMENT": lambda state, action: state + 1,
            "DECREMENT": lambda state, action: state - 1,
        })
    """
    def reducer(state: Any, action: Action) -> Any:
        action_type = action.get_type()
        
        if action_type in handlers:
            return handlers[action_type](state, action)
        
        if default:
            return default(state, action)
        
        return state
    
    return reducer


def identity_reducer(state: Any, action: Action) -> Any:
    """
    Reducer que no hace nada (retorna el mismo estado).
    
    Útil como placeholder o para testing.
    """
    return state


if __name__ == "__main__":
    from .action import SimpleAction, PayloadAction
    
    # Ejemplo 1: Reducer simple
    def counter_reducer(state: int, action: Action) -> int:
        if action.get_type() == "INCREMENT":
            return state + 1
        elif action.get_type() == "DECREMENT":
            return state - 1
        elif action.get_type() == "SET":
            return action.payload if hasattr(action, 'payload') else state
        return state
    
    # Test
    state = 0
    state = counter_reducer(state, SimpleAction("INCREMENT"))
    print(f"After INCREMENT: {state}")  # 1
    
    state = counter_reducer(state, SimpleAction("INCREMENT"))
    print(f"After INCREMENT: {state}")  # 2
    
    state = counter_reducer(state, SimpleAction("DECREMENT"))
    print(f"After DECREMENT: {state}")  # 1
    
    state = counter_reducer(state, PayloadAction("SET", payload=10))
    print(f"After SET(10): {state}")  # 10
    
    # Ejemplo 2: ReducerBuilder
    print("\n--- ReducerBuilder ---")
    
    builder_reducer = (
        ReducerBuilder(initial_state=0)
        .case("INCREMENT", lambda state, action: state + 1)
        .case("DECREMENT", lambda state, action: state - 1)
        .case("MULTIPLY", lambda state, action: state * action.payload)
        .build()
    )
    
    state = 5
    state = builder_reducer(state, SimpleAction("INCREMENT"))
    print(f"After INCREMENT: {state}")  # 6
    
    state = builder_reducer(state, PayloadAction("MULTIPLY", payload=3))
    print(f"After MULTIPLY(3): {state}")  # 18
    
    # Ejemplo 3: CombinedReducer
    print("\n--- CombinedReducer ---")
    
    def todos_reducer(state: list, action: Action) -> list:
        if state is None:
            state = []
        
        if action.get_type() == "ADD_TODO":
            return [*state, action.payload]
        elif action.get_type() == "CLEAR_TODOS":
            return []
        return state
    
    combined = CombinedReducer({
        "counter": counter_reducer,
        "todos": todos_reducer
    })
    
    app_state = {"counter": 0, "todos": []}
    
    app_state = combined.reduce(app_state, SimpleAction("INCREMENT"))
    print(f"After INCREMENT: {app_state}")
    
    app_state = combined.reduce(app_state, PayloadAction("ADD_TODO", payload="Learn Vela"))
    print(f"After ADD_TODO: {app_state}")
    
    app_state = combined.reduce(app_state, PayloadAction("ADD_TODO", payload="Build app"))
    print(f"After ADD_TODO: {app_state}")

File: packages/state-management/src/reducer.rs
/*
Implementación de Reducer functions para el sistema de Store Redux-style

Historia: VELA-035R (EPIC-03D State Management)
Tarea: TASK-035T
Fecha: 2025-12-09

Descripción:
Funciones reductoras puras que transforman el estado basado en acciones.
Implementa pattern matching y composición funcional para reducers Redux-style.
*/

use std::sync::Arc;

/// Type alias para funciones reductoras puras
/// Una función reductora toma el estado actual y una acción, y retorna el nuevo estado
pub type Reducer<State, A> = Arc<dyn Fn(&State, &A) -> State + Send + Sync>;

/// Builder para crear reducers compuestos
pub struct ReducerBuilder<State> {
    reducers: Vec<Box<dyn Fn(&State, &dyn std::any::Any) -> Option<State> + Send + Sync>>,
}

impl<State> ReducerBuilder<State>
where
    State: Clone + Send + Sync + 'static,
{
    /// Crear nuevo builder
    pub fn new() -> Self {
        ReducerBuilder {
            reducers: Vec::new(),
        }
    }

    /// Agregar un reducer para un tipo específico de acción
    pub fn add_reducer<A>(mut self, reducer_fn: impl Fn(&State, &A) -> State + Send + Sync + 'static) -> Self
    where
        A: 'static,
    {
        let boxed_reducer = Box::new(move |state: &State, action: &dyn std::any::Any| {
            action.downcast_ref::<A>().map(|a| reducer_fn(state, a))
        });
        self.reducers.push(boxed_reducer);
        self
    }

    /// Construir el reducer compuesto
    pub fn build(self) -> Reducer<State, Box<dyn std::any::Any + Send + Sync>> {
        let reducers = Arc::new(self.reducers);

        Arc::new(move |state: &State, action: &Box<dyn std::any::Any + Send + Sync>| {
            for reducer in reducers.iter() {
                if let Some(new_state) = reducer(state, action.as_ref()) {
                    return new_state;
                }
            }
            // Si ningún reducer maneja la acción, retornar estado sin cambios
            state.clone()
        })
    }
}

/// Función helper para crear reducers simples
pub fn create_reducer<State, A>(
    reducer_fn: impl Fn(&State, &A) -> State + Send + Sync + 'static,
) -> Reducer<State, A>
where
    State: 'static,
    A: 'static,
{
    Arc::new(reducer_fn)
}

/// Función helper para combinar múltiples reducers
pub fn combine_reducers<State>(
    reducers: Vec<Reducer<State, Box<dyn std::any::Any + Send + Sync>>>,
) -> Reducer<State, Box<dyn std::any::Any + Send + Sync>>
where
    State: Clone + Send + Sync + 'static,
{
    Arc::new(move |state: &State, action: &Box<dyn std::any::Any + Send + Sync>| {
        let mut current_state = state.clone();
        for reducer in &reducers {
            current_state = reducer(&current_state, action);
        }
        current_state
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::any::Any;

    #[derive(Debug, Clone, PartialEq)]
    struct CounterState {
        count: i32,
    }

    #[derive(Debug)]
    struct Increment;

    #[derive(Debug)]
    struct Decrement;

    #[derive(Debug)]
    struct SetValue {
        value: i32,
    }

    impl CounterState {
        fn new() -> Self {
            CounterState { count: 0 }
        }
    }

    #[test]
    fn test_simple_reducer() {
        let initial_state = CounterState::new();
        let reducer = create_reducer(|state: &CounterState, _action: &Increment| {
            CounterState {
                count: state.count + 1,
            }
        });

        let action = Increment;
        let new_state = reducer(&initial_state, &action);

        assert_eq!(new_state.count, 1);
    }

    #[test]
    fn test_reducer_builder() {
        let builder = ReducerBuilder::new()
            .add_reducer(|state: &CounterState, _action: &Increment| {
                CounterState {
                    count: state.count + 1,
                }
            })
            .add_reducer(|state: &CounterState, _action: &Decrement| {
                CounterState {
                    count: state.count - 1,
                }
            })
            .add_reducer(|state: &CounterState, action: &SetValue| {
                CounterState {
                    count: action.value,
                }
            });

        let reducer = builder.build();
        let initial_state = CounterState::new();

        // Test increment
        let action: Box<dyn Any + Send + Sync> = Box::new(Increment);
        let new_state = reducer(&initial_state, &action);
        assert_eq!(new_state.count, 1);

        // Test decrement
        let action: Box<dyn Any + Send + Sync> = Box::new(Decrement);
        let new_state = reducer(&new_state, &action);
        assert_eq!(new_state.count, 0);

        // Test set value
        let action: Box<dyn Any + Send + Sync> = Box::new(SetValue { value: 42 });
        let new_state = reducer(&new_state, &action);
        assert_eq!(new_state.count, 42);
    }

    #[test]
    fn test_combine_reducers() {
        // Crear reducers que manejan el mismo tipo de acción
        let increment_reducer: Reducer<CounterState, Box<dyn Any + Send + Sync>> = Arc::new(|state: &CounterState, action: &Box<dyn Any + Send + Sync>| {
            if let Some(_inc) = action.downcast_ref::<Increment>() {
                CounterState {
                    count: state.count + 1,
                }
            } else {
                state.clone()
            }
        });

        let multiply_reducer: Reducer<CounterState, Box<dyn Any + Send + Sync>> = Arc::new(|state: &CounterState, action: &Box<dyn Any + Send + Sync>| {
            if let Some(_inc) = action.downcast_ref::<Increment>() {
                CounterState {
                    count: state.count * 2,
                }
            } else {
                state.clone()
            }
        });

        let combined = combine_reducers(vec![increment_reducer, multiply_reducer]);
        let initial_state = CounterState { count: 5 };

        let action: Box<dyn Any + Send + Sync> = Box::new(Increment);
        let new_state = combined(&initial_state, &action);

        // Primero incrementa (5 + 1 = 6), luego multiplica (6 * 2 = 12)
        assert_eq!(new_state.count, 12);
    }

    #[test]
    fn test_reducer_immutability() {
        let initial_state = CounterState::new();
        let reducer = create_reducer(|state: &CounterState, _action: &Increment| {
            CounterState {
                count: state.count + 1,
            }
        });

        let action = Increment;
        let new_state = reducer(&initial_state, &action);

        // Estado original no debe cambiar
        assert_eq!(initial_state.count, 0);
        assert_eq!(new_state.count, 1);
    }
}